pkg/resource/managed_reconciler.go

```diff
@@ -0,0 +1,438 @@
+
+	// Panic early if we've been asked to reconcile a resource kind that has not
+	// been registered with our controller manager's scheme.
+	_ = nm()
+
+	r := &ManagedReconciler{
+		client:     m.GetClient(),
+		newManaged: nm,
+		shortWait:  defaultManagedShortWait,
+		longWait:   defaultManagedLongWait,
+		managed:    defaultMRManaged(m),
+		external:   defaultMRExternal(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile a managed resource with an external resource.
+func (r *ManagedReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	// NOTE(negz): This method is a little over our cyclomatic complexity goal.
+	// Be wary of adding additional complexity.
+
+	log.V(logging.Debug).Info("Reconciling", "controller", managedControllerName, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), managedReconcileTimeout)
+	defer cancel()
+
+	managed := r.newManaged()
+	if err := r.client.Get(ctx, req.NamespacedName, managed); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetManaged)
+	}
+
+	external, err := r.external.Connect(ctx, managed)
+	if err != nil {
+		// We'll usually hit this case if our Provider or its secret are missing
+		// or invalid. If this is first time we encounter this issue we'll be
+		// requeued implicitly when we update our status with the new error
+		// condition. If not, we want to try again after a short wait.
+		managed.SetConditions(v1alpha1.ReconcileError(err))
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	o, err := external.Observe(ctx, managed)
+	if err != nil {
+		// We'll usually hit this case if our Provider credentials are invalid
+		// or insufficient for observing the external resource type we're
+		// concerned with. If this is the first time we encounter this issue
+		// we'll be requeued implicitly when we update our status with the new
+		// error condition. If not, we want to try again after a short wait.
+		managed.SetConditions(v1alpha1.ReconcileError(err))
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	if err := r.managed.Publish(ctx, managed, o.ConnectionDetails); err != nil {
+		// If this is the first time we encounter this issue we'll be requeued
+		// implicitly when we update our status with the new error condition. If
+		// not, we want to try again after a short wait.
+		managed.SetConditions(v1alpha1.ReconcileError(err))
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	if meta.WasDeleted(managed) {
```

Comment:
I think I agree.

In most cases today the delete case is the first one we consider. In this controller design I had initially moved the observation phase above the delete phase because part of the observation reports whether the external resource currently exists. At the time I had a `Delete` method (that would explicitly delete the external resource, as opposed to `Finalize` which leaves whether the resource is deleted up to the implementation) and my thinking was that we could avoid calling `Delete` if we had already observed the external resource did not actually exist.

I ended up deciding against that path, and instead decided the `Finalize` implementation should be required to handle the case in which it's being asked to delete an external resource that does not actually exist, but neglected to move the delete block back up above the finalize block.

Reply:
Pushed a change that updates this behaviour.